src/openforms/submissions/tasks/registration.py

```diff
@@ -0,0 +1,119 @@
+
+    The reference is used in the payment, confirmation page and confirmation e-mail. It
+    is possible that the backend registration failed or does not generate a reference
+    (read: case number), in which case we need to generate a unique reference ourselves.
+    """
+    submission = Submission.objects.get(id=submission_id)
+    # idempotency - do not run this again if there already is a reference!
+    if submission.public_registration_reference:
+        logger.info(
+            "Submission %d already had a public registration reference, aborting.",
+            submission_id,
+        )
+        return
+
+    # race-condition detection. There's no reason/logic after the number 5 other than
+    # gut feeling.
+    MAX_NUM_ATTEMPTS = 5
+    race_condition_suspected, save_attempts = False, 0
+
+    while (
+        race_condition_suspected or save_attempts == 0
+    ) and save_attempts < MAX_NUM_ATTEMPTS:
+        try:
+            with transaction.atomic():
+                reference = get_reference_for_submission(submission)
+                submission.public_registration_reference = reference
+                submission.save(update_fields=["public_registration_reference"])
+        except IntegrityError as error:
+            race_condition_suspected = True
+            logger.warning(
+                "Likely race condition being handled for submission %d, did %d save attempts.",
+                submission_id,
+                save_attempts + 1,
+                exc_info=error,
+            )
+            save_attempts += 1
+            # if we're about to leave the loop, re-raise the original exception
+            if save_attempts >= MAX_NUM_ATTEMPTS:
+                raise
+        else:
+            return reference
+    # should never reach this
+    raise RuntimeError(  # noqa
+        "Unexpected code-path! Either the reference should have been saved successfully, "
+        "or the error should have been re-raised."
+    )
+
+
+def get_reference_for_submission(submission: Submission) -> str:
+    # Circular import
+    from openforms.registrations.service import (
+        NoSubmissionReference,
+        extract_submission_reference,
+    )
+
+    try:
+        reference = extract_submission_reference(submission)
+    except NoSubmissionReference as exc:
+        logger.info(
+            "Could not get the reference from the registration result for submission %d, "
+            "generating one instead",
+            submission.id,
+            exc_info=exc,
+        )
+        reference = generate_unique_submission_reference()
+    return reference
+
+
+def get_random_reference() -> str:
+    # 36 characters with length 6 -> 36^6 possible combinations.
+    # that's roughly 2.1 billion combinations before we run out of options.
+    # Also note that submissions are pruned after a (configurable) number of days, so
+    # used references do become available again after that time.
+    # TODO: maybe include date param?
```

Comment:
as discussed this morning - holding off on that, but we keep it in the back of our mind this may be needed in the future in the event that a reference gets recycled.